docs/proposals/client-package-structure.md

```diff
@@ -0,0 +1,153 @@
+# Client: layering and package structure
+
+## Desired layers
+
+### Transport
+
+The transport layer is concerned with round-tripping requests to an apiserver
+somewhere. It consumes a Config object with options appropriate for this.
+(That's most of the current client.Config structure.)
+
+Third parties (e.g., open shift) may have their own implementation.
+
+Transport delivers an object that implements http's RoundTripper interface
+and/or can be used in place of http.DefaultClient to route requests.
+
+Transport objects are safe for concurrent use, and are cached and reused by
+subsequent layers.
+
+Tentative name: "Transport".
+
+Action items:
+* Split out of current client package into a new package. (@krousey)
+
+### RESTClient/request.go
+
+RESTClient consumes a Transport, a group, and a version, and produces something
+that implements the interface currently in request.go. That is, with a
+RESTClient, you can write chains of calls like: c.Get().Path(p).Param("name",
+"value").Do()
+
+The group and version may be empty when constructing a RESTClient. This is valid
+for executing discovery commands. The group and version may be overridable with
+a function call.
+
+RESTClients may have semantic information built in. For example, if the
```

Comment:
Couldn't this be accomplished at the transport layer? Currently you can provide your own transport or even provide something that wraps a transport. I imagine this RESTClient could just consume a transport that does this?